README.md
# raspi_system_temperature
Measure 2 temps, Show 2 temps and using raspberry Pi 3 B+

To make sys_kama.py work as daemon(service):  
  
	1. ln -s /home/pi/src/sys_kama/sys_kama.service /etc/systemd/system/sys_kama.service  
	2. systemctl enable /etc/systemd/system/sys_kama.service  
	3. sudo reboot  
  
	Warning:  
		1. recognize path to sys_kama.sh  
		2. chmod 750 sys_kama.sh  
	How to recognize addition to service:  
		systemctl list-unit-files --type=service | grep sys_kama  
  
  history:  
  	version3.0:	今まで専属で使っていたLCDモジュールを、共用にします。  
			そのため、一時、LCDモジュールを使えなくします。  
			ライブラリとしてのACM1602は温存します。  




CHANGELOG.md
## 3.0.3 (12月 20, 2018)
  - Bump version to 3.0.2.  v3.0.2...HEAD
  - Bump version to 3.0.1.  v3.0.2...HEAD
  - ACM1602 LCDモジュールをを専有しないようにします そのために、サービスのshellもファイル名を変更しました。 いままでのsys_kama.pyは残します。 	modified:   README.md 	modified:   sys_kama.sh 	new file:   sys_kama_v3.0.0.py  v3.0.2...HEAD
  - Create new release-3.0 branch  v3.0.2...HEAD
  - Merge from the branch release-2.3  v3.0.2...HEAD
  - For what i do not know ??? 	new file:   last_path.sh  v3.0.2...HEAD
  - Bump version to 2.3.1.  v3.0.2...HEAD
  - Create new release-2.3 branch  v3.0.2...HEAD
  - Merge from the branch release-2.2  v3.0.2...HEAD
  - Append about daemon work 	modified:   README.md  v3.0.2...HEAD
  - To work as daemon 	new file:   sys_kama.service  v3.0.2...HEAD
  - Work as daemon 	new file:   sys_kama.sh  v3.0.2...HEAD
  - Bump version to 2.2.1.  v3.0.2...HEAD
  - Create new release-2.2 branch  v3.0.2...HEAD
  - Merge from the branch release-2.1  v3.0.2...HEAD
  - Add LCD to system itself 	new file:   acm1602.py 	new file:   character_table.py 	modified:   max31856.py 	modified:   sys_kama.py  v3.0.2...HEAD
  - writing now  v3.0.2...HEAD
  - add renew* in .gitignore  v3.0.2...HEAD
  - version1.3  v3.0.2...HEAD
  - Create new release-2.1 branch  v3.0.2...HEAD
  - Merge from the branch release-2.0  v3.0.2...HEAD
  - Bump version to 2.0.1.  v3.0.2...HEAD
  - Create new release-2.0 branch  v3.0.2...HEAD
  - Merge from the branch release-1.0  v3.0.2...HEAD
  - Bump version to 1.0.3.  v3.0.2...HEAD
  - Bump version to 1.0.2.  v3.0.2...HEAD
  - Bump version to 1.0.1.  v3.0.2...HEAD
  - Add VERSION and CHANGELOG.md files, Bump version to v1.0.0.  v3.0.2...HEAD
  - Modify 	modified:   CHANGELOG.md  v3.0.2...HEAD
  - Add VERSION and CHANGELOG.md files, Bump version to v0.1.0.  v3.0.2...HEAD
  - Initial commit  v3.0.2...HEAD

## 3.0.2 (12月 20, 2018)
  - Bump version to 3.0.1.  v3.0.1...HEAD
  - ACM1602 LCDモジュールをを専有しないようにします そのために、サービスのshellもファイル名を変更しました。 いままでのsys_kama.pyは残します。 	modified:   README.md 	modified:   sys_kama.sh 	new file:   sys_kama_v3.0.0.py  v3.0.1...HEAD
  - Create new release-3.0 branch  v3.0.1...HEAD
  - Merge from the branch release-2.3  v3.0.1...HEAD
  - For what i do not know ??? 	new file:   last_path.sh  v3.0.1...HEAD
  - Bump version to 2.3.1.  v3.0.1...HEAD
  - Create new release-2.3 branch  v3.0.1...HEAD
  - Merge from the branch release-2.2  v3.0.1...HEAD
  - Append about daemon work 	modified:   README.md  v3.0.1...HEAD
  - To work as daemon 	new file:   sys_kama.service  v3.0.1...HEAD
  - Work as daemon 	new file:   sys_kama.sh  v3.0.1...HEAD
  - Bump version to 2.2.1.  v3.0.1...HEAD
  - Create new release-2.2 branch  v3.0.1...HEAD
  - Merge from the branch release-2.1  v3.0.1...HEAD
  - Add LCD to system itself 	new file:   acm1602.py 	new file:   character_table.py 	modified:   max31856.py 	modified:   sys_kama.py  v3.0.1...HEAD
  - writing now  v3.0.1...HEAD
  - add renew* in .gitignore  v3.0.1...HEAD
  - version1.3  v3.0.1...HEAD
  - Create new release-2.1 branch  v3.0.1...HEAD
  - Merge from the branch release-2.0  v3.0.1...HEAD
  - Bump version to 2.0.1.  v3.0.1...HEAD
  - Create new release-2.0 branch  v3.0.1...HEAD
  - Merge from the branch release-1.0  v3.0.1...HEAD
  - Bump version to 1.0.3.  v3.0.1...HEAD
  - Bump version to 1.0.2.  v3.0.1...HEAD
  - Bump version to 1.0.1.  v3.0.1...HEAD
  - Add VERSION and CHANGELOG.md files, Bump version to v1.0.0.  v3.0.1...HEAD
  - Modify 	modified:   CHANGELOG.md  v3.0.1...HEAD
  - Add VERSION and CHANGELOG.md files, Bump version to v0.1.0.  v3.0.1...HEAD
  - Initial commit  v3.0.1...HEAD

## 3.0.1 (12月 20, 2018)
  - ACM1602 LCDモジュールをを専有しないようにします そのために、サービスのshellもファイル名を変更しました。 いままでのsys_kama.pyは残します。 	modified:   README.md 	modified:   sys_kama.sh 	new file:   sys_kama_v3.0.0.py  v3.0.0...HEAD
  - Create new release-3.0 branch  v3.0.0...HEAD
  - Merge from the branch release-2.3  v3.0.0...HEAD
  - For what i do not know ??? 	new file:   last_path.sh  v3.0.0...HEAD
  - Bump version to 2.3.1.  v3.0.0...HEAD
  - Create new release-2.3 branch  v3.0.0...HEAD
  - Merge from the branch release-2.2  v3.0.0...HEAD
  - Append about daemon work 	modified:   README.md  v3.0.0...HEAD
  - To work as daemon 	new file:   sys_kama.service  v3.0.0...HEAD
  - Work as daemon 	new file:   sys_kama.sh  v3.0.0...HEAD
  - Bump version to 2.2.1.  v3.0.0...HEAD
  - Create new release-2.2 branch  v3.0.0...HEAD
  - Merge from the branch release-2.1  v3.0.0...HEAD
  - Add LCD to system itself 	new file:   acm1602.py 	new file:   character_table.py 	modified:   max31856.py 	modified:   sys_kama.py  v3.0.0...HEAD
  - writing now  v3.0.0...HEAD
  - add renew* in .gitignore  v3.0.0...HEAD
  - version1.3  v3.0.0...HEAD
  - Create new release-2.1 branch  v3.0.0...HEAD
  - Merge from the branch release-2.0  v3.0.0...HEAD
  - Bump version to 2.0.1.  v3.0.0...HEAD
  - Create new release-2.0 branch  v3.0.0...HEAD
  - Merge from the branch release-1.0  v3.0.0...HEAD
  - Bump version to 1.0.3.  v3.0.0...HEAD
  - Bump version to 1.0.2.  v3.0.0...HEAD
  - Bump version to 1.0.1.  v3.0.0...HEAD
  - Add VERSION and CHANGELOG.md files, Bump version to v1.0.0.  v3.0.0...HEAD
  - Modify 	modified:   CHANGELOG.md  v3.0.0...HEAD
  - Add VERSION and CHANGELOG.md files, Bump version to v0.1.0.  v3.0.0...HEAD
  - Initial commit  v3.0.0...HEAD

## 3.0 (12月 20,2018)
  - Merge from the branch release-2.3  v2.3.1...HEAD
  - For what i do not know ??? 	new file:   last_path.sh  v2.3.1...HEAD
  - Bump version to 2.3.1.  v2.3.1...HEAD
  - Create new release-2.3 branch  v2.3.1...HEAD
  - Merge from the branch release-2.2  v2.3.1...HEAD
  - Append about daemon work 	modified:   README.md  v2.3.1...HEAD
  - To work as daemon 	new file:   sys_kama.service  v2.3.1...HEAD
  - Work as daemon 	new file:   sys_kama.sh  v2.3.1...HEAD
  - Bump version to 2.2.1.  v2.3.1...HEAD
  - Create new release-2.2 branch  v2.3.1...HEAD
  - Merge from the branch release-2.1  v2.3.1...HEAD

ht16k33.py
#!/usr/bin/env python3

#https://www.javadrive.jp/python/string/index9.html
#数値を文字列に変換(str)


import smbus

import time

class Ht16k33:
	DDD = { }
	DDD["0"] = 0x3F
	DDD["1"] = 0x06
	DDD["2"] = 0x5B
	DDD["3"] = 0x4F
	DDD["4"] = 0x66
	DDD["5"] = 0x6D
	DDD["6"] = 0x7D
	DDD["7"] = 0x27
	DDD["8"] = 0x7F
	DDD["9"] = 0x6F
	DDD["-"] = 0x00
	DDD["F"] = 0x71
	DDD["A"] = 0x77
	DDD["U"] = 0x3E
	DDD["L"] = 0x38

	def __init__(self,dev_bus,dev_addr):
		#(hex) dev_bus
		#(hex) dev_addr
		self.bus = smbus.SMBus(dev_bus)
		self.addr = dev_addr
		self.bus.write_byte(self.addr,0x21)
		self.bus.write_byte(self.addr,0x81)
		self.bus.write_byte(self.addr,0xE0)

	def version(self):
		return "1.1"


		#light strength
	def strength(self,value):
		#(hex) value : 0x0～0xF
		value = int(value)
		if value > 0xF:
			value = 0xF
		elif value < 0x0:
			value = 0x0
		value = value + 0xE0;
		self.bus.write_byte(self.addr,value)

	def print(self,sValue="0000",fColon=False):
		#(String) sValue
		#(boolean) fColon : 0:disappear  1:appear
		sValue = "----" + sValue
		flag = True
		try:
			d4 = Ht16k33.DDD[sValue[-1]]
			d3 = Ht16k33.DDD[sValue[-2]]
			d2 = Ht16k33.DDD[sValue[-3]]
			d1 = Ht16k33.DDD[sValue[-4]]
		except:
			print("sValue contains another of [0-9] letter")
			flag = False
		if flag:
			if fColon == True:
				coron = 0xf
			else:
				coron = 0x0
			data = [d1,0,d2,0,coron,0,d3,0,d4]
			self.bus.write_i2c_block_data(self.addr,0x00,data)

	def close(self):
		print("you don't have to close smbus")

if __name__ == "__main__":

	#SAMPLE
	#	usage:
	#		example1 : $ python3 ht16k33.py 1
	#			if you connect 1 4Dig7Sed-LED
	#		example2 : $ python3 ht16k33.py 2
	#			if you connect 2 4Dig7Sed-LEDs

	from datetime import datetime
	import sys
	
	arg = sys.argv
	if len(arg) == 1:
		print("enter number of 4dig7seg-LED device(s),expect [1-2]")
		h = Ht16k33(1,0x70)
		h.print("FAUL",2)
	else:
		if arg[1] == "1":
			flag1 = True
			flag2 = False
		elif arg[1] == "2":
			flag1 = True
			flag2 = True
		else:
			print("Wrong number,expect [1-2]")
			flag1 = False
			flag2 = False

		if flag1:
			DEVICE_BUS = 1
			DEVICE_ADDR0 = 0x70
			DEVICE_ADDR1 = 0x71

			h1 = Ht16k33(DEVICE_BUS,DEVICE_ADDR0)
			if flag2:h2 = Ht16k33(DEVICE_BUS,DEVICE_ADDR1)

			coron = 0
			count = 1 
			flag = True
			while True:
				try:
					count += 1
					if count == 2:
						ms = datetime.now().strftime("%M%S")
						mult = datetime.now().strftime("%d%H")

						count = 0
					if flag:
						flag = False
						coron = True
					else:
						flag = True
						coron = False
					h1.print(ms,coron)
					if flag2:h2.print(mult,coron)

					time.sleep(0.5)

				except KeyboardInterrupt:
					break



max31856.py
#!/usr/bin/env python3
# _*_ coding:utf-8 _*_

import sys
import time
import spidev


class Max31856:

	spi = None
	DUMMY = 0

	def __init__(self,numberOfSPI,numberOfChip):
		#B,T,R,K,...の選択などをするべきだろう
		self.spi = spidev.SpiDev()
		try:
			self.spi.open(numberOfSPI,numberOfChip)			#/sys/bus/spi/devices/dev0.0を使う
		except FileNotFoundError:
			print("指定されたSPIデバイスは接続されていません")
			self.spi = None
		else:
			self.spi.mode = 0x03			#モード3 CPOL:1 CPHA:1 ,Especially CPHA must be 1
			self.spi.max_speed_hz = 1000000	#最大クロック周波数

	def version(self):
		return "1.3"

	def analyze_fault(self,code):
		if code:
			if code & 1:
				print("断線")
			if code & 2:
				print("電圧が低すぎるか、高すぎる")
			if code & 4:
				print("温度が設定下限を超えている")
			if code & 8:
				print("温度が設定上限を超えている")
			if code & 16:
				print("冷接点温度が設定下限を超えている")
			if code & 32:
				print("冷接点温度が設定上限を超えている")
			if code & 64:
				print("冷接点温度が仕様の使用範囲を超えている")
			if code & 128:
				print("熱接点温度が仕様の使用温度を超えている")
		else:
			print("正常")

	def analyze_fault_for_LCD(self,code):
		moji = ""
		if code & 1:moji +="ダンセン　"
		if code & 2:moji += "デンアツイジョウ　"
		if code & 4:moji +="オンドカゲン　"
		if code & 8:moji +="オンドジョウゲン　"
		if code & 16:moji +="レイセッテンオンドシタスギ　"
		if code & 32:moji +="レイセッテンオンドウエスギ　"
		if code & 64:moji +="レイセッテン　シヨウハンイゴエ　"
		if code & 128:moji +="ネツセッテン　シヨウハンイゴエ"
		return moji




	def close(self):
		if self.spi is not None:
			self.spi.close()

	def reset(self):
		#FAULTピンのリセット
		resp = self.spi.xfer2([0x80,0x17])

	def read(self):
		if self.spi is None:
			print("SPI device has not been opend")
			return { "FAULT":-1000,"HJ":-1000000,"CJ":-1000000}
		else:
			dummy = self.DUMMY

			#通常計測
			#xfer2はcsを下げたまま
			#xferは1バイトごとにcsを下げ上げする
			resp = self.spi.xfer2([0x80,0x55,0x35,0x00,0x3C,0xF6,0x57,0x80,0xFE,0xC0,0x00])	#設定と同時に計測するのでCR0の第6ビットを立てる
			time.sleep(1)			#計測を待つ


			#Faultの検知
			resp = self.spi.xfer([0x0f,dummy])
			if (resp[1] & 0xFF) != 0:
				valueFlt = resp[1]
				valueHJ = -10000
				valueCJ = -10000
			else:

				valueFlt = resp[1]

				#熱接点の温度を読み込み
				resp = self.spi.xfer2([0x0C,0x0D,0x0E,dummy])
				valueHJ = resp[1] * 256 + resp[2]
				if (resp[1] & 0x80) != 0:
					valueHJ = -1 * (~(valueHJ - 1) & 0x7FFF)	#2の補数の10進数化(これに0.0625を掛けると温度)
				#冷接点の温度を読み込み
				resp = self.spi.xfer2([0x0A,0x0B,dummy])
				valueCJ = (resp[1] << 6)  + (resp[2] >> 2)
				if (resp[1] & 0x80) != 0:
					valueCJ = -1 * (~(valueCJ - 1) & 0x7FFF)	#2の補数の10進数化(これに0.015を掛けると温度)

			return self._format(valueFlt,valueHJ,valueCJ)

	def _format(self,valueFlt,valueHJ,valueCJ):
		aa = { }
		aa["FAULT"] = valueFlt
		aa["HJ"] = valueHJ
		aa["CJ"] = valueCJ
		return aa

	def write_file(filepath,jsonFormat):
		file = codecs.open(filepath,"w","utf-8")
		file.write(jsonFormat)
		file.close()

if __name__ == '__main__':


	args = sys.argv;
	flag = True
	if len(args) < 3:flag = False
	elif len(args) > 4:flag = False
	elif args[1] != "0" and args[1] != "1":flag = False
	elif args[1] == "0":
		if args[2] != "0" and args[2] != "1":flag = False
	elif args[1] == "1":
		if args[2] != "1" and args[2] != "2":flag = False

	if flag is False:
		print("Can't open Max31856")
		print("Max31856 class Usage : python3 max31856.py n1 n2 [reset]")
		print("n1...SPI NUMBER : 0 or 1")
		print("n2...CHIP NUMBER")
		print("                : 0 or 1 if SPI NUMBER is 0")
		print("                : 1 or 2 if SPI NUMBER is 1")
	else:
		spi = Max31856(int(args[1]),int(args[2]))
		if "reset" in sys.argv:
			spi.reset()
			spi.close()
		else:
			vals = spi.read()
			print("Fault {}".format(vals["FAULT"]))
			print("HJ {}".format(vals["HJ"]))
			print("CJ {}".format(vals["CJ"]))
			spi.analyze_fault()
			spi.close()




sys_kama.py
import math
import sys
sys.path.append("/home/pi/src/sys_kama")
from ht16k33 import Ht16k33
from max31856 import Max31856
from acm1602 import ACM1602
import time
import threading
	
def mes_pri(spi,ht,gyou):
	temps = spi.read()
	#print(temps)
	if temps["FAULT"] == 0:
		temp = temps["HJ"]*0.0625
		tempHT =str(math.floor(temp))
		tempACM = str(temp)
	else:
#		spi.analyze_fault(temps["FAULT"])
		tempACM = spi.analyze_fault_for_LCD(temps["FAULT"])
		tempHT = "F-"+("0"+str(temps["FAULT"]))[-2:]
	ht.print(tempHT,2)
	if gyou == 1:
		acm.line1()
	elif gyou == 2:
		acm.line2()
	acm.sendMessage(tempACM)
	time.sleep(1)


DEVICE_BUS = 1
DEVICE_ADDR0 = 0x71
DEVICE_ADDR1 = 0x70

ht0 = Ht16k33(DEVICE_BUS,DEVICE_ADDR0)
ht1 = Ht16k33(DEVICE_BUS,DEVICE_ADDR1)

SPI = 0
CE0 = 0
CE1 = 1
spi0 = Max31856(SPI,CE0)
spi1 = Max31856(SPI,CE1)

acm = ACM1602()
acm.sendMessage("FAULT")
acm.displayOnOffControl.set(acm.DisplayOnOffControl.DISPON_CURSOROFF_BLINKOFF)

while True:
	mes_pri(spi0,ht0,1)
	mes_pri(spi1,ht1,2)
	time.sleep(1)


sys_kama_v3.0.0.py
import math
import sys
sys.path.append("/home/pi/src/sys_kama")
from ht16k33 import Ht16k33
from max31856 import Max31856
#v3.0.0 from acm1602 import ACM1602
import time
import threading
	
def mes_pri(spi,ht,gyou):
	temps = spi.read()
	#print(temps)
	if temps["FAULT"] == 0:
		temp = temps["HJ"]*0.0625
		tempHT =str(math.floor(temp))
		tempACM = str(temp)
	else:
#		spi.analyze_fault(temps["FAULT"])
		tempACM = spi.analyze_fault_for_LCD(temps["FAULT"])
		tempHT = "F-"+("0"+str(temps["FAULT"]))[-2:]
	ht.print(tempHT,2)
	if gyou == 1:
#v3.0.0		acm.line1()
		pass
	elif gyou == 2:
#v3.0.0		acm.line2()
		pass
#v3.0.0	acm.sendMessage(tempACM)
	time.sleep(1)


DEVICE_BUS = 1
DEVICE_ADDR0 = 0x71
DEVICE_ADDR1 = 0x70

ht0 = Ht16k33(DEVICE_BUS,DEVICE_ADDR0)
ht1 = Ht16k33(DEVICE_BUS,DEVICE_ADDR1)

SPI = 0
CE0 = 0
CE1 = 1
spi0 = Max31856(SPI,CE0)
spi1 = Max31856(SPI,CE1)

#v3.0.0 acm = ACM1602()
#v3.0.0 acm.sendMessage("FAULT")
#v3.0.0 acm.displayOnOffControl.set(acm.DisplayOnOffControl.DISPON_CURSOROFF_BLINKOFF)

while True:
	mes_pri(spi0,ht0,1)
	mes_pri(spi1,ht1,2)
	time.sleep(1)
